extmod/moduevent_native.c

```diff
@@ -0,0 +1,244 @@
+};
+
+typedef struct _mp_obj_poll_entry_t {
+    mp_obj_poll_entry_base_t base;
+    struct _mp_obj_poll_entry_t *next;
+    mp_uint_t (*ioctl)(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode);
+    uint16_t flags_waiting;
+    uint16_t flags_ready;
+} mp_obj_poll_entry_t;
+
+STATIC mp_obj_poll_entry_t *mp_uevent_poll_entry_new(mp_obj_poll_t *poller) {
+    if (poller->len >= poller->alloc) {
+        size_t new_alloc = poller->alloc + MICROPY_ALLOC_UEVENT_POLL_INC;
+        poller->entries = m_renew(mp_obj_poll_entry_t, poller->entries, poller->alloc, new_alloc);
+        poller->alloc = new_alloc;
+    }
+    mp_obj_poll_entry_t *self = &poller->entries[poller->len++];
+    self->base.base.type = &mp_type_poll_entry;
+    self->base.poller = poller;
+    self->base.user_method_name = MP_QSTR_NULL;
+    self->base.user_data = mp_const_none;
+    self->next = NULL;
+    return self;
+}
+
+STATIC void mp_uevent_poll_entry_init(mp_obj_poll_entry_t *self, mp_obj_t obj) {
+    self->base.obj = obj;
+    self->base.user_flags = 0;
+    self->ioctl = mp_get_stream_raise(obj, MP_STREAM_OP_IOCTL)->ioctl;
+    self->flags_waiting = 0;
+    self->flags_ready = 0;
+}
+
+STATIC void mp_uevent_poll_entry_process_event(mp_obj_poll_entry_t *entry, uint16_t flags) {
+    if (!(entry->flags_waiting & flags)) {
+        return;
+    }
+    uint16_t old_flags_ready = entry->flags_ready;
+    entry->flags_ready |= entry->flags_waiting & flags;
+    entry->flags_waiting &= ~flags;
+    if (old_flags_ready == 0 && entry->flags_ready) {
+        // Put entry on the ready linked list.
+        entry->next = entry->base.poller->entry_ready;
+        entry->base.poller->entry_ready = entry;
+    }
+}
+
+STATIC void mp_uevent_entry_clear_events(mp_obj_poll_t *self, mp_obj_poll_entry_t *entry, uint16_t flags) {
+    entry->flags_waiting &= ~flags;
+    if (entry->flags_waiting == 0) {
+        // Move entry to free list.
+        entry->base.obj = mp_const_none;
+        entry->next = self->entry_free;
+        self->entry_free = entry;
+    }
+}
+
+mp_obj_t mp_uevent_poll_iternext(mp_obj_t self_in) {
+    mp_obj_poll_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (self->entry_ready != NULL) {
+        mp_obj_poll_entry_t *entry = self->entry_ready;
+        self->entry_ready = entry->next;
+        entry->base.user_flags = entry->flags_ready;
+        entry->flags_ready = 0;
+        mp_uevent_entry_clear_events(self, entry, 0);
+        return MP_OBJ_FROM_PTR(entry);
+    }
+
+    return MP_OBJ_STOP_ITERATION;
+}
+
+mp_obj_t mp_uevent_poll_register(size_t n_args, const mp_obj_t *args) {
+    mp_obj_poll_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_uint_t flags;
+    if (n_args == 3) {
+        flags = mp_obj_get_int(args[2]);
+        // flags must be non-zero.
+    } else {
+        flags = 0xffff;
+    }
+    mp_obj_poll_entry_t *entry = NULL;
+    for (size_t i = 0; i < self->len; ++i) {
+        if (self->entries[i].base.obj == args[1]) {
```

Comment:
there's an implicit assumption here that this equality test is sufficient. for file descriptors it is. But looking ahead, it assumes that if I register `Pin(12)` that this is a true singleton object such that if I call `Pin(12)` in another part of the app and also register that the equality check is sufficient. Ughh, I hope this makes sense. I don't think this is necessarily incorrect or bad code, but it might be useful to flag this assumption.

Reply:
Yes it's essentially an "is" comparison in Python language, which should be enough.  A port should ensure that `Pin(12)` here and there is the same object.

But the idea is to make this much more efficient and query the object to see if it's already registered, see if it already has its event callback set to the correct value (see #6110).  This will be a more correct test because it relates to the event itself rather than the object.